File: src/types/video.ts
/**
 * Mirror of `src-tauri/src/models/video.rs` (Spike F — Video Engine).
 * Field names stay snake_case so serde defaults line up with the wire.
 */

export type VideoEventCategory =
  | "colisao"
  | "frenagem"
  | "impacto"
  | "reacao"
  | "semaforo"
  | "mudanca_faixa"
  | "outro";

export interface VideoMedia {
  id: string;
  occurrence_id: string;
  original_path: string | null;
  /** Workspace-relative path (videos/originais/<filename>). */
  relative_path: string;
  filename: string;
  sha256: string;
  size_bytes: number;
  duration_s: number | null;
  codec: string | null;
  width: number | null;
  height: number | null;
  pixel_format: string | null;
  fps_declared: number | null;
  /** "30000/1001" — kept as the string ffprobe returned. */
  avg_frame_rate: string | null;
  r_frame_rate: string | null;
  time_base: string | null;
  frame_count: number | null;
  bitrate: number | null;
  /** ffprobe -of json verbatim. */
  raw_probe_json: string;
  /** JSON array of strings — technical warnings. */
  warnings_json: string;
  created_at: string;
  updated_at: string;
}

export interface VideoEvent {
  id: string;
  occurrence_id: string;
  media_hash: string;
  timestamp_s: number;
  timestamp_label: string;
  frame_observed: number | null;
  pts: number | null;
  time_base: string | null;
  category: string;
  title: string;
  description: string;
  reviewed: boolean;
  source: string;
  created_at: string;
  updated_at: string;
}

export interface VideoExport {
  id: string;
  occurrence_id: string;
  media_hash: string;
  event_id: string | null;
  type: string;
  requested_timestamp_s: number;
  actual_timestamp_s: number | null;
  delta_s: number | null;
  output_path: string;
  filename: string;
  sidecar_json_path: string | null;
  details_json: string;
  created_at: string;
}

export interface VideoStoryboardFrame {
  id: string;
  occurrence_id: string;
  media_hash: string;
  event_id: string | null;
  export_id: string | null;
  title: string;
  caption: string;
  notes: string;
  requested_timestamp_s: number;
  actual_timestamp_s: number | null;
  delta_s: number | null;
  observed_frame_index: number | null;
  estimated_total_frames: number | null;
  frame_index_is_estimated: boolean;
  pts: number | null;
  time_base: string | null;
  output_path: string;
  sidecar_json_path: string | null;
  reviewed: boolean;
  created_at: string;
  updated_at: string;
}

export interface VideoOperationLog {
  id: number;
  occurrence_id: string;
  media_hash: string | null;
  action: string;
  details_json: string;
  created_at: string;
}

export interface VideoBundle {
  media: VideoMedia;
  events: VideoEvent[];
  exports: VideoExport[];
  storyboard: VideoStoryboardFrame[];
}

export interface RegisterVideoInput {
  source_path: string;
}

export interface CreateVideoEventInput {
  media_hash: string;
  timestamp_s: number;
  category: VideoEventCategory | string;
  title: string;
  description?: string;
}

export interface UpdateVideoEventInput {
  title?: string;
  description?: string;
  category?: VideoEventCategory | string;
  timestamp_s?: number;
  reviewed?: boolean;
}

export interface CollectFrameInput {
  media_hash: string;
  timestamp_s: number;
  event_id?: string | null;
  title?: string;
  caption?: string;
  notes?: string;
}

export interface CollectFrameResult {
  export: VideoExport;
  storyboard_frame: VideoStoryboardFrame;
  warnings: string[];
}

export interface UpdateStoryboardFrameInput {
  title?: string;
  caption?: string;
  notes?: string;
  reviewed?: boolean;
}

File: src-tauri/src/models/video.rs
//! Video models (Spike F — Video Engine).
//!
//! Mirrors the Python lab models (`SICRO_VIDEO_LAB_RELATORIO.md` §3)
//! adapted to the SICRO Desktop wire format. Field names use snake_case;
//! the frontend types in `src/types/video.ts` mirror them 1:1.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// VideoMedia

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoMedia {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    /// Caminho original no disco do usuário (informativo).
    pub original_path: Option<String>,
    /// Caminho relativo dentro do workspace: `videos/originais/<filename>`.
    pub relative_path: String,
    pub filename: String,
    pub sha256: String,
    pub size_bytes: u64,
    pub duration_s: Option<f64>,
    pub codec: Option<String>,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub pixel_format: Option<String>,
    pub fps_declared: Option<f64>,
    /// String "30000/1001" — preservamos fidelidade técnica.
    pub avg_frame_rate: Option<String>,
    pub r_frame_rate: Option<String>,
    pub time_base: Option<String>,
    pub frame_count: Option<i64>,
    pub bitrate: Option<i64>,
    pub raw_probe_json: String,
    pub warnings_json: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RegisterVideoInput {
    /// Absolute path on the user's disk (returned by the OS file dialog).
    pub source_path: String,
}

// ---------------------------------------------------------------------------
// VideoEvent

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoEvent {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub media_hash: String,
    pub timestamp_s: f64,
    pub timestamp_label: String,
    pub frame_observed: Option<i64>,
    pub pts: Option<i64>,
    pub time_base: Option<String>,
    pub category: String,
    pub title: String,
    pub description: String,
    pub reviewed: bool,
    pub source: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateVideoEventInput {
    pub media_hash: String,
    pub timestamp_s: f64,
    pub category: String,
    pub title: String,
    #[serde(default)]
    pub description: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateVideoEventInput {
    pub title: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub timestamp_s: Option<f64>,
    pub reviewed: Option<bool>,
}

// ---------------------------------------------------------------------------
// VideoExport (PNG frame extraction)

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoExport {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub media_hash: String,
    pub event_id: Option<Uuid>,
    /// 'frame_png' — reserved discriminator for future export kinds.
    pub r#type: String,
    pub requested_timestamp_s: f64,
    pub actual_timestamp_s: Option<f64>,
    pub delta_s: Option<f64>,
    /// Workspace-relative path: `videos/storyboards/frames/<filename>.png`.
    pub output_path: String,
    pub filename: String,
    pub sidecar_json_path: Option<String>,
    /// Full extraction context (timestamp delta, ffmpeg version, etc.).
    pub details_json: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CollectFrameInput {
    pub media_hash: String,
    pub timestamp_s: f64,
    pub event_id: Option<Uuid>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub caption: Option<String>,
    #[serde(default)]
    pub notes: Option<String>,
}

/// Result of `collect_video_frame`: the export row + the storyboard row
/// already linked, so the frontend can update both panels in one go.
#[derive(Debug, Clone, Serialize)]
pub struct CollectFrameResult {
    pub export: VideoExport,
    pub storyboard_frame: VideoStoryboardFrame,
    pub warnings: Vec<String>,
}

// ---------------------------------------------------------------------------
// VideoStoryboardFrame

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoStoryboardFrame {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub media_hash: String,
    pub event_id: Option<Uuid>,
    pub export_id: Option<Uuid>,
    pub title: String,
    pub caption: String,
    pub notes: String,
    pub requested_timestamp_s: f64,
    pub actual_timestamp_s: Option<f64>,
    pub delta_s: Option<f64>,
    pub observed_frame_index: Option<i64>,
    pub estimated_total_frames: Option<i64>,
    /// True when the frame index is an estimate (FPS-based), false when
    /// FFmpeg returned a confident PTS-anchored position.
    pub frame_index_is_estimated: bool,
    pub pts: Option<i64>,
    pub time_base: Option<String>,
    pub output_path: String,
    pub sidecar_json_path: Option<String>,
    pub reviewed: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateStoryboardFrameInput {
    pub title: Option<String>,
    pub caption: Option<String>,
    pub notes: Option<String>,
    pub reviewed: Option<bool>,
}

// ---------------------------------------------------------------------------
// VideoOperationLog

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VideoOperationLog {
    pub id: i64,
    pub occurrence_id: Uuid,
    pub media_hash: Option<String>,
    pub action: String,
    pub details_json: String,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// Aggregated bundle returned by `open_video_media` so the UI can hydrate
// the whole module in a single round-trip.

#[derive(Debug, Clone, Serialize)]
pub struct VideoBundle {
    pub media: VideoMedia,
    pub events: Vec<VideoEvent>,
    pub exports: Vec<VideoExport>,
    pub storyboard: Vec<VideoStoryboardFrame>,
}
